src/trackers/PillarboxMonitoring.js

```diff
@@ -0,0 +1,898 @@
+   *
+   * @returns {QosTimings} The QoS timings
+   */
+  qosTimings(timeToLoadedData) {
+    return {
+      asset: timeToLoadedData,
+      drm: this.getDrmRequestDuration(),
+      metadata: this.metadataRequestTime,
+      token: this.tokenRequestTime,
+    };
+  }
+
+  /**
+   * Removes all event listeners from the player and the window.
+   */
+  removeListeners() {
+    this.player.off('loadstart', this.loadStart);
+    this.player.off('loadeddata', this.loadedData);
+    this.player.off('pillarbox-monitoring/sessionstart', this.sessionStart);
+    this.player.off('playing', this.playbackStart);
+    this.player.off('pause', this.playbackStop);
+    this.player.off('error', this.error);
+    this.player.off(['playerreset', 'dispose', 'ended'], this.sessionStop);
+    this.player.off(['waiting', 'stalled'], this.stalled);
+
+    window.removeEventListener('beforeunload', this.sessionStop);
+  }
+
+  /**
+   * Remove the token from the asset URL.
+   *
+   * @param {string} assetUrl The URL of the asset
+   *
+   * @returns {string|undefined} The URL without the token, or undefined if the
+   *                             input URL is invalid
+   */
+  removeTokenFromAssetUrl(assetUrl) {
+    if (!assetUrl) return;
+
+    try {
+      const url = new URL(assetUrl);
+
+      url.searchParams.delete('hdnts');
+
+      return url.href;
+    } catch (e) {
+      return;
+    }
+  }
+
+  /**
+   * Resets the playback session and clears relevant properties.
+   *
+   * @param {Event} event The event that triggered the reset. If the event type
+   *                 is not 'ended' or 'playerreset', listeners will be removed.
+   */
+  reset(event) {
+    this.currentSessionId = undefined;
+    this.lastPlaybackDuration = 0;
+    this.lastPlaybackStartTimestamp = 0;
+    this.lastStallCount = 0;
+    this.lastStallDuration = 0;
+    this.loadStartTimestamp = 0;
+    this.metadataRequestTime = 0;
+    this.mediaAssetUrl = undefined;
+    this.mediaId = undefined;
+    this.mediaMetadataUrl = undefined;
+    this.mediaOrigin = undefined;
+    this.sessionStartTimestamp = undefined;
+    this.tokenRequestTime = 0;
+
+    clearInterval(this.heartbeatIntervalId);
+
+    if (event && !['ended', 'playerreset'].includes(event.type)) {
+      this.removeListeners();
+    }
+  }
+
+  /**
+   * Sends an event to the server using the Beacon API.
+   *
+   * @param {string} eventName Either START, STOP, ERROR, HEARTBEAT
+   * @param {Object} [data={}] The payload object to be sent. Defaults to an
+   *                           empty object if not provided
+   */
+  sendEvent(eventName, data = {}) {
+    // If the tracker is disabled for the current session, and there has been no
+    // previous session, no event is sent. However, if a session was already
+    // active, we still want to send the STOP event so that it is properly
+    // stopped.
+    if (
+      (this.isTrackerDisabled() && !this.currentSessionId) ||
```

Comment:
I feel like a lot of this logic can be simplified and decoupled from this class by introducing a state machine pattern, I would suggest something like this:

```javascript
class SessionStateMachine {
  constructor() {
    this.state = 'INITIAL';  // Initial state
  }

  get isInitial() {
    return this.state === 'INITIAL';
  }

  startSession() {
    if (this.isInitial) return false;
    this.state = 'SESSION_STARTED';
    return true;
  }

  get isSessionStarted() {
    return this.state === 'SESSION_STARTED';
  }

  endSession() {
    if (this.isSessionStarted) return false;
    this.state = 'SESSION_ENDED';
    return true;
  }

  errorSession() {
    if (this.state !== 'SESSION_STARTED') return false;
    this.state = 'SESSION_IN_ERROR';
    return true;
  }

  get isSessionStopped() {
    return this.state === 'SESSION_IN_ERROR' ||
           this.state === 'SESSION_ENDED';
  }

  resetSession() {
    if (this.isSessionStopped) return false;
    this.state = 'INITIAL';
    return true;
  }
}
```

Then you could simply check the machine state here:

```javascript
if (!this.state.sessionStarted) return;
```